File: apple/Sources/R2CCore/OperatingProfiles.swift
import Foundation

/// Readable, immutable authority snapshots. None of these advisories gates recording.
public enum OperatingProfiles {
    public static var standard: [String: Any] { ["id": "standard-part-107", "version": 1, "name": "Standard Part 107 — no operational waiver", "authorityType": "part107", "conditions": [[String: Any]]()] }
    public static var bvlos: [String: Any] { ["id": "bvlos-pending", "version": 1, "name": "BVLOS — authority details pending", "authorityType": "unresolved", "conditions": [[String: Any]]()] }
    public static var other: [String: Any] { ["id": "other-pending", "version": 1, "name": "Other / details pending", "authorityType": "unresolved", "conditions": [[String: Any]]()] }
    public static func json(_ value: Any) -> String { (try? JSONSerialization.data(withJSONObject: value, options: [.sortedKeys])).map { String(decoding: $0, as: UTF8.self) } ?? "{}" }
    public static func object(_ value: String?) -> [String: Any] { (try? JSONSerialization.jsonObject(with: Data((value ?? "{}").utf8))) as? [String: Any] ?? [:] }
    public static func choices(_ state: [String: Any]) -> [[String: Any]] {
        [standard] + ((state["operatingProfiles"] as? [String: Any])?["profiles"] as? [[String: Any]] ?? []) + [bvlos, other]
    }
    public static func defaultProfile(_ state: [String: Any]) -> [String: Any] {
        let id = (state["operatingProfiles"] as? [String: Any])?["defaultProfileId"] as? String ?? "standard-part-107"
        return choices(state).first { $0["id"] as? String == id } ?? other.merging(["missingProfileId": id]) { _, new in new }
    }
    /// Restore the choice, using current catalog details rather than a previous flight snapshot.
    public static func preferredProfile(_ state: [String: Any], savedID: String?) -> [String: Any] {
        guard let id = savedID, !id.isEmpty else { return defaultProfile(state) }
        return choices(state).first { $0["id"] as? String == id }
            ?? other.merging(["missingProfileId": id]) { _, new in new }
    }
    public static func preferenceKey(organization: String) -> String {
        "operating-profile-choice:" + organization.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
    }
    public static func warnings(_ profile: [String: Any], state: [String: Any], pilotID: String, aircraftID: String, incidentID: String, managed: Bool, now: Date = Date()) -> [String] {
        var issues: [String] = []
        let authority = profile["authorityType"] as? String ?? "unresolved"
        let today = String(ISO8601DateFormatter().string(from: now).prefix(10))
        if !["part107", "part107_waiver"].contains(authority) { issues.append("Operating authority and qualification requirements need review.") }
        if profile["missingProfileId"] != nil { issues.append("Selected profile is missing; confirm authority details.") }
        for key in ["effectiveFrom", "effectiveUntil"] {
            if let value = profile[key] as? String, !value.isEmpty {
                if ISO8601DateFormatter().date(from: value + "T00:00:00Z") == nil { issues.append("Profile effective date needs review.") }
                else if (key == "effectiveFrom" && value > today) || (key == "effectiveUntil" && value < today) { issues.append("Profile is outside its effective dates.") }
            }
        }
        if authority == "part107_waiver" && ["waiverNumber", "holder", "document", "effectiveFrom", "effectiveUntil"].contains(where: { (profile[$0] as? String ?? "").isEmpty }) { issues.append("Waiver details are incomplete; review the source document.") }
        for (key, selected) in [("pilotIds", pilotID), ("aircraftIds", aircraftID), ("locationIds", incidentID)] {
            if let ids = profile[key] as? [String], !ids.isEmpty, !ids.contains(selected) { issues.append("Profile applicability needs review: \(key).") }
        }
        if managed {
            let formatter = ISO8601DateFormatter()
            formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
            let fetched = formatter.date(from: state["fetchedAt"] as? String ?? "") ?? ISO8601DateFormatter().date(from: state["fetchedAt"] as? String ?? "")
            if fetched == nil || now.timeIntervalSince(fetched!) > 86400 || now.timeIntervalSince(fetched!) < -300 { issues.append("Saved profiles are missing or stale; synchronize when available.") }
            let current = choices(state).first { $0["id"] as? String == profile["id"] as? String }
            if current == nil || current?["version"] as? Int != profile["version"] as? Int { issues.append("Profile has changed or is unavailable; review the saved version.") }
        }
        return issues
    }
    public static func snapshot(_ profile: [String: Any], state: [String: Any], pilotID: String, aircraftID: String, incidentID: String, assignmentID: String, managed: Bool, checked: Set<Int>) -> [String: Any] {
        ["profile": profile, "selectedAt": ISO8601DateFormatter().string(from: Date()), "organizationId": state["organizationId"] as? String ?? "",
         "incidentId": incidentID, "assignmentId": assignmentID, "catalogRevision": (state["operatingProfiles"] as? [String: Any])?["revision"] as? Int ?? 0,
         "fetchedAt": state["fetchedAt"] as? String ?? "", "checkedConditions": checked.sorted(),
         "reviewIssues": warnings(profile, state: state, pilotID: pilotID, aircraftID: aircraftID, incidentID: incidentID, managed: managed) + (((profile["conditions"] as? [[String: Any]] ?? []).indices.contains { !checked.contains($0) }) ? ["One or more briefing conditions have not been acknowledged."] : [])]
    }
    public static func active(_ readiness: FlightReadiness?) -> [String: Any] {
        let changes = (try? JSONSerialization.jsonObject(with: Data((readiness?.operatingProfileChangesJSON ?? "[]").utf8))) as? [[String: Any]] ?? []
        return changes.last?["after"] as? [String: Any] ?? object(readiness?.operatingProfileJSON)
    }
    public static func retainingHistory(previous: FlightReadiness?, next: FlightReadiness) -> FlightReadiness {
        guard let previous, previous.operatingProfileJSON != nil else { return next }
        if next.operatingProfileJSON == nil {
            var retained = next; retained.operatingProfileJSON = previous.operatingProfileJSON
            retained.operatingProfileChangesJSON = previous.operatingProfileChangesJSON
            return retained
        }
        var result = next
        var changes = (try? JSONSerialization.jsonObject(with: Data((previous.operatingProfileChangesJSON ?? "[]").utf8))) as? [[String: Any]] ?? []
        let old = active(previous), selected = object(next.operatingProfileJSON)
        if json(old["profile"] ?? [:]) != json(selected["profile"] ?? [:]) || json(old["checkedConditions"] ?? []) != json(selected["checkedConditions"] ?? []) {
            changes.append(["at": ISO8601DateFormatter().string(from: Date()), "before": old, "after": selected])
        }
        result.operatingProfileJSON = previous.operatingProfileJSON
        result.operatingProfileChangesJSON = json(changes)
        return result
    }
}

/// Explicit assignment session, scoped by organization and incident IDs, never persisted.
public struct OperatingProfileAssignment: Sendable {
    private var scope = ""
    public private(set) var assignmentID = ""
    public private(set) var profileJSON: String?
    public init() {}
    public mutating func setScope(organization: String, incident: String) {
        let next = OperatingProfiles.json([organization, incident])
        if scope != next { scope = next; end() }
    }
    public mutating func end() { assignmentID = ""; profileJSON = nil }
    public mutating func remember(_ profile: [String: Any]) {
        if assignmentID.isEmpty { assignmentID = UUID().uuidString }
        profileJSON = OperatingProfiles.json(profile)
    }
}

File: app/src/main/java/org/ncssar/rid2caltopo/data/OperatingProfiles.kt
package org.ncssar.rid2caltopo.data

import org.json.JSONArray
import org.json.JSONObject
import java.time.Instant
import java.time.LocalDate
import java.util.UUID

/** Profile decisions are advisory and do not participate in flight admission. */
object OperatingProfiles {
    fun standard() = JSONObject().put("id", "standard-part-107").put("version", 1)
        .put("name", "Standard Part 107 — no operational waiver").put("authorityType", "part107").put("conditions", JSONArray())
    fun bvlos() = JSONObject().put("id", "bvlos-pending").put("version", 1).put("name", "BVLOS — authority details pending").put("authorityType", "unresolved").put("conditions", JSONArray())
    fun other() = JSONObject().put("id", "other-pending").put("version", 1)
        .put("name", "Other / details pending").put("authorityType", "unresolved").put("conditions", JSONArray())
    fun choices(state: JSONObject): List<JSONObject> = listOf(standard()) +
        state.optJSONObject("operatingProfiles")?.optJSONArray("profiles").objects() + listOf(bvlos(), other())
    fun default(state: JSONObject): JSONObject {
        val id = state.optJSONObject("operatingProfiles")?.optString("defaultProfileId", "standard-part-107") ?: "standard-part-107"
        return choices(state).firstOrNull { it.optString("id") == id }
            ?: other().put("missingProfileId", id)
    }
    fun preferenceKey(organization: String) = "operating-profile-choice:" + organization.trim('/')
    fun preferred(state: JSONObject, savedId: String?): JSONObject {
        if (savedId.isNullOrBlank()) return default(state)
        return choices(state).firstOrNull { it.optString("id") == savedId }
            ?: other().put("missingProfileId", savedId)
    }
    fun warnings(profile: JSONObject, state: JSONObject, pilotId: String, aircraftId: String, mapId: String,
                 managed: Boolean, now: Instant = Instant.now()): List<String> = buildList {
        val today = now.toString().take(10)
        if (profile.optString("authorityType") !in listOf("part107", "part107_waiver")) add("Operating authority and qualification requirements need review.")
        if (profile.has("missingProfileId")) add("Selected profile is missing; confirm authority details.")
        for (key in listOf("effectiveFrom", "effectiveUntil")) {
            val date = profile.optString(key)
            if (date.isNotBlank()) {
                if (runCatching { LocalDate.parse(date) }.isFailure) add("Profile effective date needs review.")
                else if ((key == "effectiveFrom" && date > today) || (key == "effectiveUntil" && date < today)) add("Profile is outside its effective dates.")
            }
        }
        if (profile.optString("authorityType") == "part107_waiver" &&
            listOf("waiverNumber", "holder", "document", "effectiveFrom", "effectiveUntil").any { profile.optString(it).isBlank() }) add("Waiver details are incomplete; review the source document.")
        for ((key, selected) in listOf("pilotIds" to pilotId, "aircraftIds" to aircraftId, "locationIds" to mapId)) {
            val ids = profile.optJSONArray(key)
            if (ids != null && ids.length() > 0 && (0 until ids.length()).none { ids.optString(it) == selected }) add("Profile applicability needs review: $key.")
        }
        if (managed) {
            val fetched = runCatching { Instant.parse(state.optString("fetchedAt")) }.getOrNull()
            if (fetched == null || fetched.isBefore(now.minusSeconds(86400)) || fetched.isAfter(now.plusSeconds(300))) add("Saved profiles are missing or stale; synchronize when available.")
            val current = choices(state).firstOrNull { it.optString("id") == profile.optString("id") }
            if (current == null || current.optInt("version") != profile.optInt("version")) add("Profile has changed or is unavailable; review the saved version.")
        }
    }
    fun snapshot(profile: JSONObject, state: JSONObject, pilotId: String, aircraftId: String, mapId: String,
                 managed: Boolean, checked: Set<Int> = emptySet()): JSONObject = JSONObject()
        .put("profile", JSONObject(profile.toString())).put("selectedAt", Instant.now().toString())
        .put("organizationId", state.optString("organizationId")).put("incidentId", mapId)
        .put("assignmentId", assignmentId).put("catalogRevision", state.optJSONObject("operatingProfiles")?.optInt("revision") ?: 0)
        .put("fetchedAt", state.optString("fetchedAt")).put("checkedConditions", JSONArray(checked.sorted()))
        .put("reviewIssues", JSONArray(warnings(profile, state, pilotId, aircraftId, mapId, managed) +
            if ((0 until (profile.optJSONArray("conditions")?.length() ?: 0)).any { it !in checked }) listOf("One or more briefing conditions have not been acknowledged.") else emptyList<String>()))

    // Assignment identity and its snapshot are session-only; the separate saved choice is not an assignment.
    private var scope = ""
    var assignmentId = ""; private set
    private var remembered: String? = null
    @JvmStatic @Synchronized fun setScope(org: String, incident: String) {
        val next = JSONArray(listOf(org, incident)).toString()
        if (next != scope) { scope = next; endAssignment() }
    }
    @JvmStatic @Synchronized fun endAssignment() { assignmentId = ""; remembered = null }
    @Synchronized fun remember(profile: JSONObject) {
        if (assignmentId.isEmpty()) assignmentId = UUID.randomUUID().toString()
        remembered = profile.toString()
    }
    @Synchronized fun initial(state: JSONObject, savedId: String? = null): JSONObject = remembered?.let { JSONObject(it) } ?: preferred(state, savedId)
    fun active(readiness: JSONObject): JSONObject? = readiness.optJSONArray("operatingProfileChanges")?.objects()?.lastOrNull()?.optJSONObject("after")
        ?: readiness.optJSONObject("operatingProfile")
    @JvmStatic fun mergeHistory(previous: String, next: String): String = runCatching {
        val before = JSONObject(previous); val after = JSONObject(next)
        val old = active(before); val selected = after.optJSONObject("operatingProfile")
        if (old != null && selected != null) {
            val changes = JSONArray(before.optJSONArray("operatingProfileChanges")?.toString() ?: "[]")
            if ((old.optJSONObject("profile")?.toString() ?: "{}") != (selected.optJSONObject("profile")?.toString() ?: "{}") ||
                (old.optJSONArray("checkedConditions")?.toString() ?: "[]") != (selected.optJSONArray("checkedConditions")?.toString() ?: "[]")) {
                changes.put(JSONObject().put("at", Instant.now().toString()).put("before", old).put("after", selected))
            }
            after.put("operatingProfile", before.getJSONObject("operatingProfile"))
            after.put("operatingProfileChanges", changes)
        }
        if (old != null && selected == null) {
            after.put("operatingProfile", before.optJSONObject("operatingProfile"))
            after.put("operatingProfileChanges", before.optJSONArray("operatingProfileChanges") ?: JSONArray())
        }
        after.toString()
    }.getOrDefault(next)
    fun JSONArray?.objects(): List<JSONObject> = (0 until (this?.length() ?: 0)).mapNotNull { this?.optJSONObject(it) }
}
